code/iaas/cluster-logic/src/main/java/io/cattle/iaas/cluster/service/impl/ClusterManagerImpl.java

```diff
@@ -0,0 +1,210 @@
+    }
+
+    private String getUri(Host cluster) {
+        return getUri(cluster, getManagingHost(cluster));
+    }
+
+    private String getConnectionPrefix(Host cluster, Host managingHost) {
+        return objectManager.isKind(managingHost, "sim") ? "sim:" : "delegate:";
+    }
+
+    private String getImagePrefix(Host cluster, Host managingHost) {
+        return objectManager.isKind(managingHost, "sim") ? "sim:" : "docker:";
+    }
+
+    private Long getNetworkIds(Host managingHost) {
+        List<? extends Network> accountNetworks = ntwkDao.getNetworksForAccount(managingHost.getAccountId(),
+                NetworkConstants.KIND_HOSTONLY);
+        if (accountNetworks.isEmpty()) {
+            // pass system network if account doesn't own any
+            List<? extends Network> systemNetworks = ntwkDao.getNetworksForAccount(accountDao.getSystemAccount()
+                    .getId(),
+                    NetworkConstants.KIND_HOSTONLY);
+            if (systemNetworks.isEmpty()) {
+                throw new RuntimeException(
+                        "Unable to find a network to start cluster server");
+            }
+            return systemNetworks.get(0).getId();
+        }
+
+        return accountNetworks.get(0).getId();
+    }
+
+    @Override
+    public Instance getClusterServerInstance(Host cluster) {
+        Agent clusterServerAgent = getClusterServerAgent(cluster);
+        Instance clusterServerInstance = null;
+        if (clusterServerAgent != null) {
+            clusterServerInstance = agentInstanceDao.getInstanceByAgent(clusterServerAgent);
+        }
+        return clusterServerInstance;
+    }
+
+    @Override
+    public Agent getClusterServerAgent(Host cluster) {
+        String uri = getUri(cluster);
+        Agent clusterServerAgent = agentInstanceDao.getAgentByUri(uri);
+        return clusterServerAgent;
+    }
+
+    @Override
+    public IpAddress getClusterServerInstanceIp(Instance clusterServerInstance) {
+        IpAddress ip = null;
+        for (Nic nic : objectManager.children(clusterServerInstance, Nic.class)) {
+            ip = ipAddressDao.getPrimaryIpAddress(nic);
+            if (ip != null) {
+                break;
+            }
+        }
+        return ip;
+    }
+
+    private void start(final Instance agentInstance) {
+        if (InstanceConstants.STATE_STOPPED.equals(agentInstance.getState())) {
+            DeferredUtils.nest(new Callable<Object>() {
+                @Override
+                public Object call() throws Exception {
+                    processManager.executeProcess(InstanceConstants.PROCESS_START, agentInstance, null);
+                    return null;
+                }
+            });
+        }
+    }
+
+    @Override
+    public Host getManagingHost(Host cluster) {
+        Long managingHostId = DataAccessor.fields(cluster).withKey(ClusterConstants.MANAGING_HOST).as(Long.class);
+        if (managingHostId == null) {
```

Comment:
If managing host can be set to null at any point, or the referenced managing host gets removed, we should handle it in the ProcessHandler. Perhaps by assigning another managing host to it?